### PR TITLE
Re-apply lost commits from legacy ocean repo

### PR DESCRIPTION
**Pull request by mihails-strasuns-sociomantic**

### CODE REVIEW

**Comment by mihails-strasuns-sociomantic** [comment]
I haven't yet checked which parts affect API and may need to be moved to v4.x.x - just picked old commits and fixes conflicts/tests.

**Comment by mathias-lang-sociomantic** [comment]
First 2 commits LGTM, last one needs a bit of adjustment.

File: src/ocean/text/convert/Memory.d [hunk 1]
```diff
@@ -87,7 +87,7 @@ public char[] memoryToHexAscii ( void[] mem, char[] output = null )
                 c = '.';
             ascii[idx] = c;
         }
-        Format.format(output, "{}\n", ascii[0 .. idx]);
```
Comment (mathias-lang-sociomantic) [inline src/ocean/text/convert/Memory.d]: You can revert this whole file, it's deprecated (https://github.com/sociomantic-tsunami/ocean/pull/174)

**Comment by mathias-lang-sociomantic** [comment]
Ping

**Comment by nemanja-boric-sociomantic** [comment]
D2 build is also red.

**Comment by mathias-lang-sociomantic** [comment]
> D2 build is also red.

Yeah, reverting changes to `ocean.text.convert.Memory` should fix it.

**Comment by mihails-strasuns-sociomantic** [comment]
Fixed.

But after re-examining actual patches I don't think we can merge it into v3.x.x anymore :( Those are risky changes that intended to be in 3.0.0 and that train is gone.

Instead I would propose to re-purpose most of this PR against v4.x.x and aim for releasing v4.0.0 roughly at the same time as v2.x.x goes out of support. What do you thing @sociomantic-tsunami/core-team @mathias-lang-sociomantic ?

**Comment by mathias-lang-sociomantic** [comment]
> But after re-examining actual patches I don't think we can merge it into v3.x.x anymore

I think first and last commit can definitely go in. The second would need some adjustment, like turning `Snoop` removal into a deprecation and reverting the changes that affect the API (`AppStatus`).
Now that `""` and `null` are treated the same I don't think anything will break (famous last words).

**Comment by mihails-strasuns-sociomantic** [comment]
> Now that "" and null are treated the same I don't think anything will break (famous last words)

That is pretty much the reason I am worried to just let it go in minor release - these last words are just _too famous_ ;)

**Comment by mihails-strasuns-sociomantic** [comment]
Have split https://github.com/sociomantic-tsunami/ocean/pull/203 into separate commit and turned snoop into deprecation.

**Comment by mihails-strasuns-sociomantic** [comment]
Ping @sociomantic-tsunami/core-team

File: src/ocean/io/console/StructTable.d [hunk 2]
```diff
@@ -274,7 +274,7 @@ public class StructTable ( S )
     private char[] defaultFieldString ( T ) ( T* field )
     {
         this.format_buffer.length = 0;
-        Format.format(this.format_buffer, "{}", *field);
+        sformat(this.format_buffer, "{}", *field);
```
Comment (mathias-lang-sociomantic) [inline src/ocean/io/console/StructTable.d]: Missing `enableStomping` here as well ?

File: src/ocean/io/select/client/TimerSet.d [hunk 3]
```diff
@@ -233,7 +232,7 @@ public class TimerSet ( EventData ) : TimerEventTimeoutManager
             protected cstring id ( )
             {
                 this.id_buf.length = 0;
```
Comment (mathias-lang-sociomantic) [inline src/ocean/io/select/client/TimerSet.d:234]: Ditto `enableStomping` ?

File: src/ocean/text/util/DigitGrouping.d [hunk 4]
```diff
@@ -278,9 +277,12 @@ public class BitGrouping
         output.length = 0;
         enableStomping(output);
 
+        if (unit.length == 0)
+            unit = "";
```
Comment (mathias-lang-sociomantic) [inline src/ocean/text/util/DigitGrouping.d]: I don't think that's necessary anymore

**Review by leandro-lucarella-sociomantic — APPROVED** [review]
I wanted to give my LGTM yesterday just when the :unicorn: madness was unleashed.

File: src/ocean/io/console/StructTable.d (hunk 2 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/io/console/StructTable.d]: ping @mihails-strasuns-sociomantic ?

File: src/ocean/io/select/client/TimerSet.d (hunk 3 above)
Comment (nemanja-boric-sociomantic) [reply src/ocean/io/select/client/TimerSet.d:234]: ping here as well.

**Comment by mihails-strasuns-sociomantic** [comment]
Updated:
- more `enableStomping`
- removed unnecessary null array special case
- rebased

**Review by mathias-lang-sociomantic — APPROVED** [review]
LGTM

**Review by nemanja-boric-sociomantic — APPROVED** [review]
LGTM

**Pull request merged**